### ISSUE TITLE
Problem with Trained Ternary Quantization

### ISSUE THREAD

**Issue by minhson** [tensorpack/tensorpack#854]
Hi. I am a newbie in Tensorflow and Tensorpack.
I am trying to reproduce Trained Ternary Quantization with ResNet and Cifar10. The program run normally. However, i cannot see Wn, Wp when it is running. What's wrong here?

This is my code:
`class Model(ModelDesc):
    def __init__(self, n):
        super(Model, self).__init__()
        self.n = n

    def inputs(self):
        return [tf.placeholder(tf.float32, [None, 32, 32, 3], 'input'),
                tf.placeholder(tf.int32, [None], 'label')
               ]

    def build_graph(self, image, label):
        #FIXMEimage, label = input_vars
        image = image / 128.0 - 1
        #FIXMEimage = image / 128.0
        #FIXMEassert tf.test.is_gpu_available()
        #FIXMEimage = tf.transpose(image, [0, 3, 1, 2])

        old_get_variable = tf.get_variable
        def new_get_variable(name, shape=None, **kwargs):
            v = old_get_variable(name, shape, **kwargs)
            # don't binarize first and last layer
            if name is not 'W' or 'conv0' in v.op.name or 'fct' in v.op.name:
                return v
            else:
                logger.info("Ternaring weight {}".format(v.op.name))
                return tw_ternarize(v, args.t)
        tf.get_variable = new_get_variable

        def conv(name, l, channel, stride):
            return Conv2D(name, l, channel, 3, stride=stride,
                          nl=tf.identity, use_bias=False,
                          W_init=tf.random_normal_initializer(stddev=np.sqrt(2.0/9/channel)))

        def residual(name, l, increase_dim=False, first=False):
            shape = l.get_shape().as_list()
            in_channel = shape[3]

            if increase_dim:
                out_channel = in_channel * 2
                stride1 = 2
            else:
                out_channel = in_channel
                stride1 = 1

            with tf.variable_scope(name) as scope:
                if not first:
                    b1 = BatchNorm('bn1', l)
                    b1 = tf.nn.relu(b1)
                else:
                    b1 = l
                c1 = conv('conv1', b1, out_channel, stride1)
                b2 = BatchNorm('bn2', c1)
                b2 = tf.nn.relu(b2)
                c2 = conv('conv2', b2, out_channel, 1)

                if increase_dim:
                    l = AvgPooling('pool', l, 2)
                    l = tf.pad(l, [[0,0], [0,0], [0,0], [in_channel//2, in_channel//2]])

                l = c2 + l
                return l

        l = conv('conv0', image, 16, 1)
        l = BatchNorm('bn0', l)
        l = tf.nn.relu(l)
        l = residual('res1.0', l, first=True)
        for k in range(1, self.n):
            l = residual('res1.{}'.format(k), l)
        # 32,c=16

        l = residual('res2.0', l, increase_dim=True)
        for k in range(1, self.n):
            l = residual('res2.{}'.format(k), l)
        # 16,c=32

        l = residual('res3.0', l, increase_dim=True)
        for k in range(1, self.n):
            l = residual('res3.' + str(k), l)
        l = BatchNorm('bnlast', l)
        l = tf.nn.relu(l)
        # 8,c=64
        l = GlobalAvgPooling('gap', l)
        logits = FullyConnected('linear', l, out_dim=10, nl=tf.identity)

        tf.get_variable = old_get_variable
        prob = tf.nn.softmax(logits, name='output')

        cost = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=label)
        cost = tf.reduce_mean(cost, name='cross_entropy_loss')

        #FIXMEwrong = prediction_incorrect(logits, label)
        #FIXMEnr_wrong = tf.reduce_sum(wrong, name='wrong')
        wrong = tf.to_float(tf.logical_not(tf.nn.in_top_k(logits, label, 1)), name='wrong_vector')

        # monitor training error
        add_moving_summary(tf.reduce_mean(wrong, name='train_error'))

        # weight decay on all W of fc layers
        wd_w = tf.train.exponential_decay(0.0002, get_global_step_var(),
                                          480000, 0.2, True)
        wd_cost = tf.multiply(wd_w, regularize_cost('.*/W', tf.nn.l2_loss), name='wd_cost')
        add_moving_summary(cost, wd_cost)

        add_param_summary(('.*/W', ['histogram']))   # monitor W

        writer = tf.summary.FileWriter('log', graph=tf.get_default_graph())
        self.cost = tf.add_n([cost, wd_cost], name='cost')

    def optimizer(self):
        #FIXMElr = tf.get_variable('learning_rate', initializer=2e-4, trainable=False)
        lr = tf.Variable(0.1, trainable=False, name='learning_rate')
        tf.summary.scalar('learning_rate', lr)
        return tf.train.MomentumOptimizer(lr, 0.9)

def get_data(train_or_test):
    isTrain = train_or_test == 'train'
    ds = dataset.Cifar10(train_or_test)
    pp_mean = ds.get_per_pixel_mean()
    if isTrain:
        augmentors = [
            imgaug.CenterPaste((40, 40)),
            imgaug.RandomCrop((32, 32)),
            imgaug.Flip(horiz=True),
            #imgaug.Brightness(20),
            #imgaug.Contrast((0.6,1.4)),
            imgaug.MapImage(lambda x: x - pp_mean),
        ]
    else:
        augmentors = [
            imgaug.MapImage(lambda x: x - pp_mean)
        ]
    ds = AugmentImageComponent(ds, augmentors)
    ds = BatchData(ds, BATCH_SIZE, remainder=not isTrain)
    if isTrain:
        ds = PrefetchData(ds, 3, 2)
    return ds

def get_config(nsize):
    mod = sys.modules['__main__']
    basename = os.path.basename(mod.__file__)
    logdir = os.path.join('train_log',
                         'ter-cifar10-resnet-{}'.format(6*nsize+2),
                         'tw', 't-{}'.format(args.t))

    logger.set_logger_dir(logdir)

    import shutil
    # import pdb; pdb.set_trace()

    shutil.copy(mod.__file__, logger.LOG_DIR)

    # prepare dataset
    dataset_train = get_data('train')
    step_per_epoch = dataset_train.size()
    dataset_test = get_data('test')

    #FIXMEsess_config = get_default_sess_config(0.9)

    get_global_step_var()
    #FIXMElr = tf.Variable(0.1, trainable=False, name='learning_rate')
    #FIXMEtf.summary.scalar('learning_rate', lr)

    #FIXMEoptimizer = tf.train.MomentumOptimizer(lr, 0.9)

    return TrainConfig(
        dataflow=dataset_train,
        #FIXMEoptimizer=optimizer,
        callbacks=[
            ModelSaver(),
            InferenceRunner(dataset_test,
                [ScalarStats('cost'), ClassificationError('wrong_vector')]),
            # PruneRunner(),
            ScheduledHyperParamSetter('learning_rate',
                                      [(1, 0.1), (82, 0.01), (123, 0.001), (150, 0.0001)])
        ],
        #FIXMEsession_config=sess_config,
        model=Model(n=NUM_UNITS),
        steps_per_epoch=step_per_epoch,
        max_epoch=400
    )

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--gpu', help='comma separated list of GPU(s) to use.') # nargs='*' in multi mode
    parser.add_argument('--load', help='load model')
    parser.add_argument('-n', '--num_units', help='Nsize', type=int, default=3)
    parser.add_argument('--t', help='hyperparameter t', type=float, default=0.05)
    args = parser.parse_args()
    NUM_UNITS = args.num_units

    if args.gpu:
        os.environ['CUDA_VISIBLE_DEVICES'] = args.gpu

    config = get_config(args.num_units)

    if args.load:
        #FIXMEassert args.load.endswith('.npz')
        #FIXMEconfig.session_init = DictRestore(dict(np.load(args.load)))
        #FIXMEconfig.session_init = SaverRestore(args.load)
        config.session_init = get_model_loader(args.load)

    #FIXMEnum_gpu = max(get_num_gpu(), 1)
    launch_train_with_config(config, SimpleTrainer())
    #FIXMElaunch_train_with_config(config, SyncMultiGPUTrainerParameterServer(num_gpu))
    #FIXMEif args.gpu:
    #FIXME    config.nr_tower = len(args.gpu.split(','))
    #FIXMESyncMultiGPUTrainer(config).train()
`
Thanks you

**Comment by ppwwyyxx** [comment]
What's inside `build_graph` is your own TensorFlow code and has nothing to do with tensorpack. As a result we don't provide support on how to implement a TensorFlow model.

Despite of that, you can refer to tensorpack examples at http://dorefa.net for a correct implementation of TTQ.
